app/grandchallenge/direct_messages/models.py

```diff
@@ -0,0 +1,199 @@
+from urllib.parse import urlencode
+
+from django.contrib.auth import get_user_model
+from django.db import models
+from django.db.models import (
+    BooleanField,
+    Case,
+    Count,
+    OuterRef,
+    Prefetch,
+    Q,
+    Subquery,
+    Value,
+    When,
+)
+from guardian.models import GroupObjectPermissionBase, UserObjectPermissionBase
+from guardian.shortcuts import assign_perm
+
+from grandchallenge.core.models import UUIDModel
+from grandchallenge.subdomains.utils import reverse
+
+
+class DirectMessage(UUIDModel):
+    conversation = models.ForeignKey(
+        "Conversation",
+        related_name="direct_messages",
+        null=False,
+        on_delete=models.CASCADE,
+    )
+    sender = models.ForeignKey(
+        get_user_model(),
+        related_name="sent_direct_messages",
+        null=True,
+        on_delete=models.SET_NULL,
+    )
+    unread_by = models.ManyToManyField(
+        get_user_model(),
+        related_name="unread_direct_messages",
+        through="DirectMessageUnreadBy",
+    )
+
+    is_reported_as_spam = models.BooleanField(default=False)
+    is_deleted = models.BooleanField(default=False)
+
+    message = models.TextField()
+
+    def get_absolute_url(self):
+        return reverse(
+            "direct-messages:conversation-detail",
+            kwargs={"pk": self.conversation.pk},
+        )
+
+    def save(self, *args, **kwargs):
+        adding = self._state.adding
+
+        super().save(*args, **kwargs)
+
+        if self.is_deleted or self.is_reported_as_spam:
+            DirectMessageUnreadBy.objects.filter(direct_message=self).delete()
+
+        if adding:
+            self.assign_permissions()
+
+    def delete(self, *args, **kwargs):
+        self.is_deleted = True
+        self.save()
```

Comment:
Why are we only removing the many to many relationship and keep the message itself?

Reply:
The unread markers are removed as there is no message to read. The message is kept but the content not displayed so that users know that a message was sent but deleted. We keep the content so we can check for abuse and also the user can un-do the deletion.

Reply:
Ok, makes sense!